frontend/catalyst/pennylane_extensions.py

```diff
@@ -168,54 +187,108 @@ class QJITDevice(qml.QubitDevice):
-        "IsingYY",
-        "IsingXY",
-        "IsingZZ",
-        "ControlledPhaseShift",
-        "CRX",
-        "CRY",
-        "CRZ",
-        "CRot",
-        "CSWAP",
-        "MultiRZ",
-        "QubitUnitary",
-        "Adjoint",
-    ]
-    observables = [
-        "Identity",
-        "PauliX",
-        "PauliY",
-        "PauliZ",
-        "Hadamard",
-        "Hermitian",
-        "Hamiltonian",
-    ]
+
+    # These must be present even if empty.
+    operations = []
+    observables = []
+
+    @staticmethod
+    def _get_operations_available_in_QIR_abstract_machine():
+        # These are the methods implemented in the runtime.
+        # A device might have gates which are native to the device
+        # but are not exposed. This is a limitation of the current runtime
+        # because it does not contain implementation of gates for every possible
+        # gate from the infinite set.
+        return {
+            "Identity",
+            "PauliX",
+            "PauliY",
+            "PauliZ",
+            "Hadamard",
+            "S",
+            "T",
+            "PhaseShift",
+            "RX",
+            "RY",
+            "RZ",
+            "Rot",
+            "CNOT",
+            "CY",
+            "CZ",
+            "SWAP",
+            "IsingXX",
+            "IsingYY",
+            "IsingXY",
+            "ControlledPhaseShift",
+            "CRX",
+            "CRY",
+            "CRZ",
+            "CRot",
+            "CSWAP",
+            "Toffoli",
+            "MultiRZ",
+            "QubitUnitary",
+        }
+
+    @staticmethod
+    def _get_operations_to_convert_to_matrix(_config):
+        # We currently override and only set MultiControlledX to preserve current behaviour.
+        # We could choose to read from config and use the "matrix" gates.
+        # However, that affects differentiability.
+        # None of the "matrix" gates with more than 2 qubits parameters are differentiable.
+        # TODO: https://github.com/PennyLaneAI/catalyst/issues/398
+        return {"MultiControlledX"}
+
+    @staticmethod
+    def _check_mid_circuit_measurement(config):
+        return config["compilation"]["mid_circuit_measurement"]
+
+    @staticmethod
+    def _check_adjoint(config):
+        return config["compilation"]["quantum_adjoint"]
+
+    @staticmethod
+    def _check_quantum_control(config):
+        return config["compilation"]["quantum_control"]
+
+    @staticmethod
+    def _set_supported_operations(config):
+        """Override the set of supported operations."""
+        native_gates = set(config["operators"]["gates"][0]["native"])
+        qir_gates = QJITDevice._get_operations_available_in_QIR_abstract_machine()
+        QJITDevice.operations = list(native_gates.intersection(qir_gates))
+
+        # These are added unconditionally.
+        QJITDevice.operations += ["Cond", "WhileLoop", "ForLoop"]
+
+        if QJITDevice._check_mid_circuit_measurement(config):  # pragma: no branch
```

Comment:
Does the braket device not cover the else branch? Or are we not running coverage with the braket tests I suppose?

Reply:
No coverage for braket.